tests/unit/test_events.py

```diff
@@ -0,0 +1,125 @@
+import re
+
+import pytest
+
+from dbt_common.events import types
+from dbt_common.events.base_types import msg_from_base_event
+from dbt_common.events.base_types import (
+    BaseEvent,
+    DebugLevel,
+    DynamicLevel,
+    ErrorLevel,
+    InfoLevel,
+    TestLevel,
+    WarnLevel,
+)
+from dbt_common.events.functions import msg_to_dict, msg_to_json
+
+
+# takes in a class and finds any subclasses for it
+def get_all_subclasses(cls):
+    all_subclasses = []
+    for subclass in cls.__subclasses__():
+        if subclass not in [TestLevel, DebugLevel, WarnLevel, InfoLevel, ErrorLevel, DynamicLevel]:
+            all_subclasses.append(subclass)
+        all_subclasses.extend(get_all_subclasses(subclass))
+    return set(all_subclasses)
+
+
+class TestEventCodes:
+    # checks to see if event codes are duplicated to keep codes singluar and clear.
+    # also checks that event codes follow correct namming convention ex. E001
+    def test_event_codes(self):
+        all_concrete = get_all_subclasses(BaseEvent)
+        all_codes = set()
+
+        for event_cls in all_concrete:
+            code = event_cls.code(event_cls)
+            # must be in the form 1 capital letter, 3 digits
+            assert re.match("^[A-Z][0-9]{3}", code)
+            # cannot have been used already
+            assert (
+                code not in all_codes
+            ), f"{code} is assigned more than once. Check types.py for duplicates."
+            all_codes.add(code)
+
+
+sample_values = [
+    # N.B. Events instantiated here include the module prefix in order to
+    # avoid having the entire list twice in the code.
+    # M - Deps generation ======================
+    types.RetryExternalCall(attempt=0, max=0),
+    types.RecordRetryException(exc=""),
+    # Z - misc ======================
+    types.SystemCouldNotWrite(path="", reason="", exc=""),
+    types.SystemExecutingCmd(cmd=[""]),
+    types.SystemStdOut(bmsg=str(b"")),
+    types.SystemStdErr(bmsg=str(b"")),
+    types.SystemReportReturnCode(returncode=0),
+    types.Formatting(),
+    types.Note(msg="This is a note."),
+]
+
+
+class TestEventJSONSerialization:
+    # attempts to test that every event is serializable to json.
+    # event types that take `Any` are not possible to test in this way since some will serialize
+    # just fine and others won't.
```

Comment:
super nit: these docstrings are great but could be standardized to the format you use below in test_bad_serialization: https://github.com/dbt-labs/dbt-common/pull/51/files#diff-018e6871938b17be4afa923d354aad1059ace4f5d0c88eceaa29711e62171289R111